include/systems/generic_projector.h

```diff
@@ -2634,13 +2649,13 @@ GenericProjector<FFunctor, GFunctor, FValue, ProjectionAction>::SubProjector::co
   for (unsigned int qp=0; qp<n_qp; qp++)
     {
       // solution at the quadrature point
-      FValue fineval = f.eval_at_point(context,
+      auto fineval = f.eval_at_point(context,
```

Comment:
After much staring I find I'm still confused.  The FValue -> auto switch here is because FValue is still being set as Number, where as f's return value is now Gradient?  If that's the case then we definitely shouldn't still be using the name "FValue"; that's just anti-self-documenting.

What I think I'd prefer is to change FValue to fit the return value type of f, then use TensorTools::MakeNumber to get the other (DoFValue?) type that in the non-vector-case is synonymous with FValue.

Reply:
> After much staring I find I'm still confused. The FValue -> auto switch here is because FValue is still being set as Number, where as f's return value is now Gradient? If that's the case then we definitely shouldn't still be using the name "FValue"; that's just anti-self-documenting.
> 
> What I think I'd prefer is to change FValue to fit the return value type of f, then use TensorTools::MakeNumber to get the other (DoFValue?) type that in the non-vector-case is synonymous with FValue.

Do we have documented what the `F` in `FFunctor` and `G` in `GFunctor` mean? I'm supplying `OldSolutionValue` as a functor. My best guess is that `F` should stand for finite element solution (value) *function* and `G` shold stand for finite element solution *gradient*? Or are `G` and `F` just supposed to stand for generic `f(x)`, `g(x)`?

Reply:
Function and Gradient, I think, but honestly this dates back to when I first added C1 element support, and I have no personal recollection at this point.

Reply:
> What I think I'd prefer is to change FValue to fit the return value type of f, then use TensorTools::MakeNumber to get the other (DoFValue?) type that in the non-vector-case is synonymous with FValue.

This is fine with me. However, I'll probably add another helper struct in `TensorTools` for determining `DofValue` since `MakeNumber<Gradient>` would just return `Gradient`, e.g. it maintains the rank of the incoming structure (vector -> vector, tensor -> tensor). Wait...there should be no reason for an additional `DofValue` type...that type should always just be `Number`...

Reply:
> However, I'll probably add another helper struct in TensorTools for determining DofValue since MakeNumber<Gradient> would just return Gradient

Yeah, I'm an idiot; I forgot that was the "upgrade Real->Number" function; I *thought* I had a "DecrementRank until you can't Decrement it no more" function but I guess not.

> Wait...there should be no reason for an additional DofValue type...that type should always just be Number...

Nope.  About a third of the reason for doing the GenericProjector refactoring in the first place was to let it handle more complicated things than Number, like the DynamicSparseNumberArray case in projection_matrix().

Reply:
It seems unfortunate that when doing projection with `OldSolutionCoefs` as the functor that `FValue` does not correspond to the `Output` template argument which I think helped lead to my confusion. But we have decided...`FValue` should correspond to the return type of `eval_at_node`/`eval_at_point`, and not to the first template argument to `FFunctor` which corresponds to the `Output` parameter provided to `FEMContext::point_value`. Correct?

Reply:
<pre>
template <typename Output,
          void (FEMContext::*point_output) (unsigned int,
                                            const Point &,
                                            Output &,
                                            const Real) const>
class OldSolutionCoefs : public OldSolutionBase<Output, point_output>
{
public:
  <b>typedef typename DSNAOutput&ltOutput&gt::type DSNA;</b>
</pre>

I like what you did here with the `DSNAOutput struct`

Reply:
I think that's correct.  FValue ought to correspond to the output of f, and eval_at_* are just ways of getting that value.  But FFunctor doesn't even technically have to be a templated class, much less be templated around an argument matching FValue.